pkg/platform/runtime/buildplan/buildplan.go

```diff
@@ -0,0 +1,286 @@
+// dependencies have been processed, it adds the artifact to the result map.
+//
+// Each artifact has a list of dependencies which also have a list of dependencies. When we
+// iterate through the artifact's dependencies, we also have to build up the dependencies of
+// each of those dependencies. Once we have a complete list of dependencies for the artifact,
+// we can continue to build up the results map.
+func buildMap(baseID strfmt.UUID, lookup map[strfmt.UUID]interface{}, result artifact.Map) error {
+	target := lookup[baseID]
+	currentArtifact, ok := target.(*model.Artifact)
+	if !ok {
+		return errs.New("Incorrect target type for id %s", baseID)
+	}
+
+	if currentArtifact.Status != model.ArtifactSucceeded {
+		return errs.New("Artifact %s did not succeed with status: %s", currentArtifact.TargetID, currentArtifact.Status)
+	}
+
+	deps := make(map[strfmt.UUID]struct{})
+	for _, depID := range currentArtifact.RuntimeDependencies {
+		deps[depID] = struct{}{}
+		recursiveDeps, err := BuildRuntimeDependencies(depID, lookup, deps)
+		if err != nil {
+			return errs.Wrap(err, "Could not build runtime dependencies for artifact %s", currentArtifact.TargetID)
+		}
+		for id := range recursiveDeps {
+			deps[id] = struct{}{}
+		}
+
+		err = buildMap(depID, lookup, result)
+		if err != nil {
+			return errs.New("Could not build map for artifact %s", currentArtifact.TargetID)
+		}
+	}
+
+	var uniqueDeps []strfmt.UUID
+	for id := range deps {
+		if _, ok := deps[id]; !ok {
+			continue
+		}
+		uniqueDeps = append(uniqueDeps, id)
+	}
+
+	info, err := GetSourceInfo(currentArtifact.GeneratedBy, lookup)
+	if err != nil {
+		return errs.Wrap(err, "Could not resolve source information")
+	}
+
+	result[strfmt.UUID(currentArtifact.TargetID)] = artifact.Artifact{
+		ArtifactID:       strfmt.UUID(currentArtifact.TargetID),
+		Name:             info.Name,
+		Namespace:        info.Namespace,
+		Version:          &info.Version,
+		RequestedByOrder: true,
+		GeneratedBy:      currentArtifact.GeneratedBy,
+		Dependencies:     uniqueDeps,
+	}
+
+	return nil
+}
+
+// SourceInfo contains useful information about the source that generated an artifact.
+type SourceInfo struct {
+	Name      string
+	Namespace string
+	Version   string
+}
+
+// GetSourceInfo retrieves the source information for an artifact. It expects the ID of the
+// source that generated the artifact and a lookup table that contains all of the sources
+// and steps in the build plan. We are able to retrieve the source information by looking
+// at the generatedBy field of the artifact and then looking at the inputs of the step that
+// generated the artifact. The inputs of the step will contain a reference to the source
+// that generated the artifact.
+//
+// The relationship is as follows:
+//
+//	Artifact (GeneratedBy) -> Step (Input) -> Source
+func GetSourceInfo(sourceID strfmt.UUID, lookup map[strfmt.UUID]interface{}) (SourceInfo, error) {
```

Comment:
Should be private as it's only used here and uses a lookup table that is only accessible from here.